src/main/java/maksym/fedorenko/bookstore/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,96 @@
+import maksym.fedorenko.bookstore.dto.shoppingcart.CreateCartItemRequestDto;
+import maksym.fedorenko.bookstore.dto.shoppingcart.UpdateCartItemRequestDto;
+import maksym.fedorenko.bookstore.exception.EntityNotFoundException;
+import maksym.fedorenko.bookstore.mapper.CartItemMapper;
+import maksym.fedorenko.bookstore.mapper.ShoppingCartMapper;
+import maksym.fedorenko.bookstore.model.Book;
+import maksym.fedorenko.bookstore.model.CartItem;
+import maksym.fedorenko.bookstore.model.ShoppingCart;
+import maksym.fedorenko.bookstore.repository.BookRepository;
+import maksym.fedorenko.bookstore.repository.CartItemRepository;
+import maksym.fedorenko.bookstore.repository.ShoppingCartRepository;
+import maksym.fedorenko.bookstore.service.ShoppingCartService;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final BookRepository bookRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
+    public CartDto getUserCart(Authentication authentication) {
+        String username = authentication.getName();
+        ShoppingCart cart = shoppingCartRepository.findByUserEmailWithCartItems(username)
+                .orElseThrow(() -> getNotFoundCartException(username));
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    public CartItemDto addCartItem(
+            Authentication authentication, CreateCartItemRequestDto requestDto) {
+        ShoppingCart cart = retrieveUserCartFromDb(authentication.getName());
+        CartItem cartItem = cartItemRepository
+                .findByShoppingCartAndBookId(cart, requestDto.bookId())
+                .orElseGet(() -> createCartItemFromRequest(requestDto, cart));
+        if (cartItem.getId() != null) {
+            cartItem.setQuantity(cartItem.getQuantity() + requestDto.quantity());
+        }
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    private CartItem createCartItemFromRequest(
+            CreateCartItemRequestDto requestDto, ShoppingCart cart) {
+        CartItem cartItem = cartItemMapper.toCartItem(requestDto);
+        cartItem.setShoppingCart(cart);
+        cartItem.setBook(getBookFromCartItem(requestDto));
+        return cartItem;
+    }
+
+    @Override
+    public CartItemDto updateCartItem(Long id, UpdateCartItemRequestDto requestDto) {
+        CartItem cartItem = findCartItemById(id);
+        cartItem.setQuantity(requestDto.quantity());
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void deleteCartItem(Long id) {
+        cartItemRepository.delete(findCartItemById(id));
```

Comment:
here's a thing you need to think of

if all the items in the cart are deleted, then the cart itself should be deleted, right?

correct me if I'm wrong

Reply:
I think that if we delete all items cart just will be empty. It's weird for me delete cart after removing all items. With empty cart we just can't make order, but we have access to cart. So I'll leave it